### PR TITLE
specify a separate `get_clustering_root_data_dir()` - handle cases where raw ephys and clustering results are stored at different root locations (e.g. different mount points)

### LINKED ISSUE #14: Waveform and ClusterQualityMetrics need masters tables
`Waveform` and `ClusterQualityMetrics` are populated with granularity of `CuratedClustering` but the primary key is with granularity of a `Unit`. This means that individual units can be deleted without triggering repopulation.

### PR DESCRIPTION
**Pull request by ttngu207**
PR includes:

+ specify a separate `get_clustering_root_data_dir()` - handle cases where raw ephys and clustering results are stored at different root locations (e.g. different mount points)
+ smaller batch ingest for Waveform and LFP - help with memory issue
+ Redesign Waveform and QualityControl as master table - fix #14

### CODE REVIEW

File: element_array_ephys/ephys.py [hunk 1]
```diff
@@ -556,75 +581,93 @@ def make(self, key):
...
+                                **unit_dict,
+                                'peak_chn_waveform_mean': channel_waveform.mean(axis=0)}
+
+                    yield unit_peak_waveform, unit_electrode_waveforms
+
+        # insert waveform on a per-unit basis to mitigate potential memory issue
+        self.insert1(key)
+        for unit_peak_waveform, unit_electrode_waveforms in yield_unit_waveforms():
+            self.Unit.insert1(unit_peak_waveform, ignore_extra_fields=True)
+            self.UnitElectrode.insert(unit_electrode_waveforms, ignore_extra_fields=True)
```
Comment (shenshan) [inline element_array_ephys/ephys.py:639]: Yield approach pretty cool!

File: element_array_ephys/ephys.py [hunk 2]
```diff
@@ -556,75 +581,93 @@ def make(self, key):
...
-    amplitude_cutoff: float  # Estimate of miss rate based on amplitude histogram
-    isolation_distance=null: float  # Distance to nearest cluster in Mahalanobis space
-    l_ratio=null: float  # 
-    d_prime=null: float  # Classification accuracy based on LDA
-    nn_hit_rate=null: float  # 
-    nn_miss_rate=null: float
-    silhouette_score=null: float  # Standard metric for cluster overlap
-    max_drift=null: float  # Maximum change in spike depth throughout recording
-    cumulative_drift=null: float  # Cumulative change in spike depth throughout recording 
+class QualityControl(dj.Imported):
```
Comment (shenshan) [inline element_array_ephys/ephys.py:646]: ClusteringQualityControl?

File: element_array_ephys/ephys.py [hunk 3]
```diff
@@ -556,75 +581,93 @@ def make(self, key):
...
+    definition = """  # Quality control metrics for a set of CuratedClustering
+    -> CuratedClustering
     """
 
-    @property
-    def key_source(self):
-        return Clustering
+    class ClusterQualityMetrics(dj.Part):
```
Comment (shenshan) [inline element_array_ephys/ephys.py:651]: UnitQualityMetrics

File: element_array_ephys/ephys.py [hunk 4]
```diff
@@ -556,75 +581,93 @@ def make(self, key):
...
+        amplitude_cutoff: float  # Estimate of miss rate based on amplitude histogram
+        isolation_distance=null: float  # Distance to nearest cluster in Mahalanobis space
+        l_ratio=null: float  # 
+        d_prime=null: float  # Classification accuracy based on LDA
+        nn_hit_rate=null: float  # 
+        nn_miss_rate=null: float
+        silhouette_score=null: float  # Standard metric for cluster overlap
+        max_drift=null: float  # Maximum change in spike depth throughout recording
+        cumulative_drift=null: float  # Cumulative change in spike depth throughout recording 
+        """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:670]: Are these quality metrics specific for ks2? To make it more general, we may consider doing something like:
```
class UnitQualityMetrics(dj.Part):
definition = """
-> master
-> CuratedClustering.Unit
"""
class UnitQualityMetric(dj.Part):
definition = """
-> master.UnitQualityMetrics
metric_name:  varchar(32)
---
metric_value=null: float
"""
If these metrics are universal, it's better to spell them out.

**Review by dimitri-yatsenko — APPROVED** [review]

**Pull request merged**